amethyst_input/src/bindings.rs

```diff
@@ -47,10 +49,90 @@ where
     pub(super) actions: HashMap<AC, SmallVec<[SmallVec<[Button; 2]>; 4]>>,
 }
 
+/// An enum of possible errors that can occur when binding an action or axis.
+#[derive(Debug, Clone, PartialEq)]
+pub enum BindingError<AX, AC> {
+    /// Combo provided for action binding has two (or more) of the same button.
+    ComboContainsDuplicates,
+    /// Combo provided was already bound to contained action.
+    ComboAlreadyBound(AC),
+    /// A combo of length 1 was provided, and it overlaps with an axis binding.
+    ButtonBoundToAxis(AX, Axis),
+    /// Axis buttons provided have overlap with an existing axis
+    AxisButtonAlreadyBoundToAxis(AX, Axis),
+    /// Axis buttons have overlap with an action combo of length 1.
+    AxisButtonAlreadyBoundToAction(AC, Button),
+    /// That specific axis on that specific controller is already in use for an
+    /// axis binding.
+    ControllerAxisAlreadyBound(AX),
+}
+
+impl<AX, AC> Display for BindingError<AX, AC> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match *self {
+            BindingError::ComboContainsDuplicates => write!(
+                f,
+                "Combo provided contained two (or more) of the same button"
+            ),
+            BindingError::ComboAlreadyBound(ref _action) => {
+                write!(f, "Combo provided was already bound to another action")
+            }
+            BindingError::ButtonBoundToAxis(ref _id, ref _axis) => {
+                write!(f, "Button provided was a button in use by an axis")
+            }
+            BindingError::AxisButtonAlreadyBoundToAxis(ref _id, ref _axis) => write!(
+                f,
+                "Axis provided contained a button that's already in use by another axis"
+            ),
+            BindingError::AxisButtonAlreadyBoundToAction(ref _id, ref _action) => write!(
+                f,
+                "Axis provided contained a button that's already in use by a single button action"
+            ),
+            BindingError::ControllerAxisAlreadyBound(ref _id) => {
+                write!(f, "Controller axis provided is already in use")
+            }
+        }
+    }
+}
+
+impl<AX, AC> Error for BindingError<AX, AC>
+where
+    AX: Debug,
+    AC: Debug,
+{
+}
+
+/// An enum of possible errors that can occur when removing an action binding.
+#[derive(Debug, Clone, PartialEq)]
+pub enum ActionRemovedError {
+    /// The action has bindings. but this isn't one of them.
+    ActionExistsButBindingDoesnt,
+    /// The action provided doesn't exist.
+    ActionMissing,
```

Comment:
Bikeshedding, could we use `ActionNotFound`? It pops up to me as consistency with generic software-library-error names

Reply:
Sure!